File: targets/python-web/vulns/vuln04_stored_xss.py
from flask import Blueprint, request, Response, jsonify
bp = Blueprint("vuln04", __name__)
_store = {}

@bp.post("/vuln04/save")
def save():
    _store[request.form.get("id", "")] = request.form.get("bio", "")
    return jsonify({"status": "saved"})

@bp.get("/vuln04/show")
def show():
    bio = _store.get(request.args.get("id", ""), "")
    # VULN:VULN-04:CWE-79:taint stored XSS: persisted bio rendered unescaped
    return Response("<html><body><div>" + bio + "</div></body></html>", mimetype="text/html")

File: targets/jsts-web/src/vulns/vuln04_stored_xss.ts
import { Router } from "express";
export const router = Router();
const store: Record<string, string> = {};
router.post("/vuln04/save", (req, res) => { store[String(req.body.id)] = String(req.body.bio || ""); res.json({ status: "saved" }); });
router.get("/vuln04/show", (req, res) => {
  const bio = store[String(req.query.id)] || "";
  // VULN:VULN-04:CWE-79:taint stored XSS: persisted bio rendered unescaped
  res.type("html").send("<html><body><div>" + bio + "</div></body></html>");
});
